### ISSUE TITLE
Add node groups (subgraphs) for VisualShader

### ISSUE THREAD

**Issue by Geometror** [godotengine/godot-proposals#7730]
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

Complex visual shader graphs can quickly become challenging to manage and maintain.
The users often need to repeat sections of their graphs, leading to redundancy and reduced readability.

### Describe the feature/enhancement and how it helps to overcome the problem or limitation

The ability to create your own nodes (as group nodes) inside of the VisualShader editor would address these issues by allowing users to encapsulate frequently used sections of a shader and reuse them across the graph/other shaders.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Some impulses for the implementation:
- `NodeGroup` is a resource containing the subgraph and information about input/output ports.
- A 'NodeGroup' node then uses this resource and can be placed in a VisuaShader graph.
- The group nodes contain a button (+shortcut/context menu action) to edit the subgraph.
- Creating such a group should be simple and retaining, e.g. just selecting the nodes it should contain -> button/context menu/shortcut -> the selected nodes are replaced automatically with the group node (which already has the correct input/output ports added)
- Subgraphs are edited in their own GraphEdit instance which replaces the currently visible instance in the VisualShader editor (may increase memory usage, but could be more performant)
- The subgraph contains at least two nodes: Group Input and Group Output. With these the input/output ports of the node group node can be configured (names/types).

I would propose an implementation very similar to Blender.

The group node:
![grafik](https://github.com/godotengine/godot-proposals/assets/50084500/aeba2638-80e9-4415-a47e-7c5a199bf0d8)
Editing the subgraph:
![grafik](https://github.com/godotengine/godot-proposals/assets/50084500/7afe8f3f-d928-4383-a0ae-a0bbeb5714e0)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

This needs to be deeply integrated into the VisualShader editor and is therefore core.

**Comment by Zaskek** [comment]
Fully supporting this;node based interfaces are basically the de facto standard in most artistic pipelines and more so in tech-art, not having the option to create subgraphs is a HUUUGE pain when using visualshaders when creating something more complex or that needs to be re-used on a lot of materials (think about a basic rain ripples-wetness effect that you would want to apply on all materials when it's raining,with a subgraph that would be as easy as creating it one time and then adding in on all materials with the hypothetic IsRaining parameter to switch it on and blend it on top of the base material). 
It would also aid non coders artists in using godot, which i think are found mainly in the middle to big sized teams/studios and could be a problem for them in choosing godot as a tech solution for their projects(?) and well we are experiencing quite the influx of them...

**Comment by dsitnick** [comment]
This should be high priority for VisualShaders.

It's a super crucial tool not just for complex shaders, but in having any kind of global functionality.
Want to make use of a singular toon ramp across multiple shaders? Forget it
Also with since the expression node doesn't support referencing an external file, it's entirely nonviable to make use of a single expression across multiple VisualShaders.

I'm sure this is already on people's radar, but it's a significant roadblock for getting familiar with these workflows.

**Comment by oxplay2** [comment]
Very good proposal

Just now, when i need add "repeatable" node-areas for a very big node graph, i could just use groups.

So what i did now was create VisualShaderNodeCustom that require "additional code" to setup it.

This feature would be very good, because i could just setup nodes in group instead of writing custom VisualShaderNodeCustom.

**Comment by ClintochX** [comment]
have they started working on this? 😢 I really need this

**Comment by Calinou** [comment]
> have they started working on this? 😢 I really need this

Not yet, but https://github.com/godotengine/godot/pull/88014 is the first step towards making this possible.

**Comment by omid3098** [comment]
With a HUGE thumbs up for this proposal, I also want to suggest having 3 different grouping methods.
- Groups: Basic grouping functionality to pack nodes and save space and preserving input and output pins. 
- Local SubGraphs: Same as groups but you can instance them and modifying one instance, changes them all. Equivalent to how we use private functions in programming languages.
- SubGraph Resource: Same as Local Subgraphs but they are saved as a separate asset and can be shared between shaders.

**Comment by ArseniyMirniy** [comment]
I really need this for my [Godot 4 Color Correction and Screen Effects Visual Shaders](https://github.com/ArseniyMirniy/Godot-4-Color-Correction-and-Screen-Effects). Frames are a good start, but we definitely require more for big graphs to avoid building the same constructions many times on a limited frame.

**Comment by Geometror** [comment]
@ArseniyMirniy Yes, I've just opened https://github.com/godotengine/godot/pull/99404. While it still requires lots of work, it's a start :)

**Comment by michaelharmonart** [comment]
> With a HUGE thumbs up for this proposal, I also want to suggest having 3 different grouping methods.
> 
>     * Groups: Basic grouping functionality to pack nodes and save space and preserving input and output pins.
> 
>     * Local SubGraphs: Same as groups but you can instance them and modifying one instance, changes them all. Equivalent to how we use private functions in programming languages.
> 
>     * SubGraph Resource: Same as Local Subgraphs but they are saved as a separate asset and can be shared between shaders.

Agreed. This would cover all the use cases. Quick grouping of nodes for organization is great, but the ability to make re-usable "functions" is HUGE. This is one of the best parts of the node groups in Blender or HDAs in Houdini. Having the ability to save as a resource and re-use between networks is also a big deal. 100% support this
